pkg/relayer/proxy/relay_meter.go

```diff
@@ -0,0 +1,384 @@
+	appMetrics, ok := supplierApps[appAddress]
+
+	// If the application is seen for the first time in this session, calculate the
+	// max amount of stake the application can consume.
+	if !ok {
+		var app apptypes.Application
+		app, err := rmtr.applicationQuerier.GetApplication(ctx, appAddress)
+		if err != nil {
+			return nil, err
+		}
+
+		sharedParams, err := rmtr.sharedQuerier.GetParams(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		service, err := rmtr.serviceQuerier.GetService(ctx, reqMeta.SessionHeader.ServiceId)
+		if err != nil {
+			return nil, err
+		}
+
+		serviceRelayDifficulty, err := rmtr.serviceQuerier.GetServiceRelayDifficulty(ctx, service.Id)
+		if err != nil {
+			return nil, err
+		}
+
+		if !createMetric {
+			return nil, ErrRelayerProxyUnknownApplication.Wrap("required metric not found")
+		}
+
+		// calculate the max amount of stake the application can consume in the current session.
+		supplierAppStake := getSupplierAppStake(app.Stake, sharedParams)
+		maxAmount := supplierAppStake.Add(rmtr.overServicingAllowanceCoins)
+		appMetrics = &appRelayMeter{
+			app:                    app,
+			consumedCoin:           cosmostypes.NewInt64Coin(volatile.DenomuPOKT, 0),
+			maxCoin:                maxAmount,
+			sessionHeader:          reqMeta.SessionHeader,
+			sharedParams:           sharedParams,
+			service:                &service,
+			serviceRelayDifficulty: serviceRelayDifficulty,
+		}
+
+		rmtr.supplierToAppMetricsMap[supplierAddress][appAddress] = appMetrics
+	}
+
+	return appMetrics, nil
+}
+
+// filterTypedEvents filters the provided events bytes for the typed event T.
+// T is then filtered by the provided filter function.
+func filterTypedEvents[T proto.Message](
+	ctx context.Context,
+	eventBzObs client.EventsBytesObservable,
+	filterFn func(T) bool,
+) observable.Observable[T] {
+	eventObs, eventCh := channel.NewObservable[T]()
+	channel.ForEach(ctx, eventBzObs, func(ctx context.Context, maybeTxBz either.Bytes) {
+		if maybeTxBz.IsError() {
+			return
+		}
+		txBz, _ := maybeTxBz.ValueOrError()
+
+		// Try to deserialize the provided bytes into an abci.TxResult.
+		txResult, err := tx.UnmarshalTxResult(txBz)
+		if err != nil {
+			return
+		}
+
+		for _, event := range txResult.Result.Events {
+			eventApplicationStakedType := cosmostypes.MsgTypeURL(*new(T))
+			if strings.Trim(event.GetType(), "/") != strings.Trim(eventApplicationStakedType, "/") {
+				continue
+			}
+
+			typedEvent, err := cosmostypes.ParseTypedEvent(event)
+			if err != nil {
+				return
+			}
+
+			castedEvent, ok := typedEvent.(T)
+			if !ok {
+				return
+			}
+
+			// Apply the filter function to the typed event.
+			if filterFn == nil || filterFn(castedEvent) {
+				eventCh <- castedEvent
+				return
+			}
+		}
+	})
+
+	return eventObs
+}
+
+// getMinedRelayCostCoin returns the cost of a relay based on the shared parameters and the service.
+// relayCost = CUPR * CUTTM * relayDifficultyMultiplier
+func getMinedRelayCostCoin(
```

Comment:
1. Cool function.
2. Do we not need/have something like this onchain?

Reply:
It's the first time we use the formula to calculate a single mined relay cost.